### DISCUSSION TITLE
installation error: Error compiling Cython file

### DISCUSSION THREAD

**Discussion by hui-nhgri** [aertslab/scenicplus#301] [Q&A]
Hi,

I'm trying to install scenicplus, but got the following error:

Using cached tables-3.8.0.tar.gz (8.0 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [58 lines of output]
      <string>:19: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
      
      Error compiling Cython file:
      ------------------------------------------------------------
      ...
      
      
      def _dump_h5_backtrace():
          cdef object bt = []
      
          if H5Ewalk(H5E_DEFAULT, H5E_WALK_DOWNWARD, e_walk_cb, <void*>bt) < 0:
                                                     ^
      ------------------------------------------------------------
      
      tables/utilsextension.pyx:384:47: Cannot assign type 'herr_t (unsigned int, const H5E_error_t *, void *) except? -1 nogil' to 'H5E_walk_t'. Exception values are incompatible. Suggest adding 'noexcept' to type 'herr_t (unsigned int, const H5E_error_t *, void *) except? -1 nogil'.
      cpuinfo failed, assuming no CPU features: 'flags'
      * Using Python 3.8.18 | packaged by conda-forge | (default, Dec 23 2023, 17:25:47)
      * Found cython 3.0.8
      * USE_PKGCONFIG: False
      * Found conda env: ``/Users/wangh31/miniconda3/envs/scenicplus``
      * Found HDF5 headers at ``/opt/homebrew/opt/hdf5/include``, library at ``/opt/homebrew/opt/hdf5/lib``.
      .. WARNING:: Could not find the HDF5 runtime.
         The HDF5 shared library was *not* found in the default library
         paths. In case of runtime problems, please remember to install it.
      * Found LZO 2 headers at ``/opt/homebrew/opt/lzo/include``, library at ``/opt/homebrew/opt/lzo/lib``.
      .. WARNING:: Could not find the LZO 2 runtime.
         The LZO 2 shared library was *not* found in the default library
         paths. In case of runtime problems, please remember to install it.
      * Skipping detection of LZO 1 since LZO 2 has already been found.
      * Found bzip2 headers at ``/Users/wangh31/miniconda3/envs/scenicplus/include``, library at ``/Users/wangh31/miniconda3/envs/scenicplus/lib``.
      * Found blosc headers at ``/opt/homebrew/opt/c-blosc/include``, library at ``/opt/homebrew/opt/c-blosc/lib``.
      .. WARNING:: Could not find the blosc runtime.
         The blosc shared library was *not* found in the default library
         paths. In case of runtime problems, please remember to install it.
      * Found blosc2 headers at ``/private/var/folders/mx/knc7yrdj3vqb5wc15967dvxmlp_nhz/T/pip-build-env-_xfuorxh/overlay/include``, library at ``/private/var/folders/mx/knc7yrdj3vqb5wc15967dvxmlp_nhz/T/pip-build-env-_xfuorxh/overlay/lib``.
        * Copying blosc2 runtime library to 'tables' dir because it was not found in standard locations
      Compiling tables/utilsextension.pyx because it changed.
      [1/1] Cythonizing tables/utilsextension.pyx
      Traceback (most recent call last):
        File "/Users/wangh31/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/wangh31/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/Users/wangh31/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/private/var/folders/mx/knc7yrdj3vqb5wc15967dvxmlp_nhz/T/pip-build-env-_xfuorxh/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 325, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
        File "/private/var/folders/mx/knc7yrdj3vqb5wc15967dvxmlp_nhz/T/pip-build-env-_xfuorxh/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 295, in _get_build_requires
          self.run_setup()
        File "/private/var/folders/mx/knc7yrdj3vqb5wc15967dvxmlp_nhz/T/pip-build-env-_xfuorxh/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 311, in run_setup
          exec(code, locals())
        File "<string>", line 928, in <module>
        File "<string>", line 923, in get_cython_extfiles
        File "/private/var/folders/mx/knc7yrdj3vqb5wc15967dvxmlp_nhz/T/pip-build-env-_xfuorxh/overlay/lib/python3.8/site-packages/Cython/Build/Dependencies.py", line 1154, in cythonize
          cythonize_one(*args)
        File "/private/var/folders/mx/knc7yrdj3vqb5wc15967dvxmlp_nhz/T/pip-build-env-_xfuorxh/overlay/lib/python3.8/site-packages/Cython/Build/Dependencies.py", line 1321, in cythonize_one
          raise CompileError(None, pyx_file)
      Cython.Compiler.Errors.CompileError: tables/utilsextension.pyx
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

In the beginning, I got the same error (https://github.com/aertslab/scatac_fragment_tools/issues/1). Then I saw some people suggest installing from the source without building a wheel. I run the code 'pip install -e . --no-binary :all:', then it started to show the above error. I cleared the environment and re-install it with the code 'pip install -e .', but it still shows the above error.

My cmake:
(scenicplus) HG-02224870-LM:scenicplus wangh31$ cmake --version
cmake version 3.28.3

My computer info:
macos: 12.7.3
machine: arm64
platform: macOS-12.7.3-arm64-arm-64bit

Could you help me take a look at this? Thank you so much! I'm new to computer science, if you need more information to find where the problem might be, please let me know. Thank you again!

**Comment by SeppeDeWinter** [comment]
Hi @hui-nhgri 

This does not seem to be an issue related to scatac_fragment_tools (or SCENIC+), but with the installation of pytables.

This discussion thread seems to be related to your issue: https://stackoverflow.com/questions/76780968/pytables-install-with-python-3-11-fails-on-macos-m1 .

I have to note that SCENIC+ has not been tested on MacOS, so there might be other installation errors.

We strongly recommend to run the tool on a Linux system in a high performance computing environment.

I hope this helps,

All the best.
Seppe

**Reply by hui-nhgri** [reply]
Hi Seppe,

I see. Thank you so much for your reply!

Best,
Hui
